### PR TITLE
feat(counting_sort): support counting_sort in c/go

### PR DESCRIPTION
**Pull request by Reanon**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/c/chapter_sorting/counting_sort.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    // 1. 统计数组最大元素 m
+    int m = 0;
+    for (int i = 0; i < size; i++) {
+        if (nums[i] > m) {
+            m = nums[i];
+        }
+    }
+    // 2. 统计各数字的出现次数
+    // counter[num] 代表 num 的出现次数
+    int *counter = malloc(sizeof(int) * size);
```
Comment (Gonglja) [inline codes/c/chapter_sorting/counting_sort.c:20]: Here the return value of malloc is `void *`, which needs to be converted to `int *`.

`int *counter =(int *) malloc(sizeof(int) * size);`

File: codes/c/chapter_sorting/counting_sort.c [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
```
Comment (Gonglja) [inline codes/c/chapter_sorting/counting_sort.c:46]: the same as above

File: codes/c/chapter_sorting/counting_sort.c [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        counter[nums[i]]++;
+    }
+    // 3. 求 counter 的前缀和，将“出现次数”转换为“尾索引”
+    // 即 counter[num]-1 是 num 在 res 中最后一次出现的索引
+    for (int i = 0; i < m; i++) {
+        counter[i + 1] += counter[i];
+    }
+    // 4. 倒序遍历 nums ，将各元素填入结果数组 res
+    // 初始化数组 res 用于记录结果
+    int *res = malloc(sizeof(int) * size);
```
Comment (Gonglja) [inline codes/c/chapter_sorting/counting_sort.c:57]: the same as above

File: codes/c/chapter_sorting/counting_sort.c (hunk 1 above)
Comment (Reanon) [reply codes/c/chapter_sorting/counting_sort.c:20]: It is unnecessary, as void * is automatically and safely promoted to any other pointer type in C.more in https://stackoverflow.com/questions/605845/do-i-cast-the-result-of-malloc

Comment (Gonglja) [reply codes/c/chapter_sorting/counting_sort.c:20]: > It is unnecessary, as void * is automatically and safely promoted to any other pointer type in C.more in https://stackoverflow.com/questions/605845/do-i-cast-the-result-of-malloc

ok, thanks @Reanon 

I took a look, there is no problem in C (sorry, I just used the g++ compiler)

Comment (Reanon) [reply codes/c/chapter_sorting/counting_sort.c:20]: 🤝

**Pull request merged**